### DISCUSSION TITLE
adding a special namespace for easy contributing

### DISCUSSION THREAD

**Discussion by younesStrittmatter** [AutoResearch/autora#362] [Ideas]
We could add a special namespace (for example, contrib) to autora (so src->autora->contrib), under which we could allow for easier contribution than in the src->autora namespace. That way, we can have contributors that don't want to bother that much with documentation or other details and quickly publish under the autora-contrib namespace (we would still have some form of qualitiy control but less strict, for example we wouldn't require tests, certain aspects of naming conventions, documentation..., and in the core-documentation, we would maybe only provid a link to these contib packages). If contributors (or we) want to contribute to the autora namespace directly, (src->autora) namespace, we would have a much higher review process.

**Comment by musslick** [comment]
That sounds interesting. How does it work with jsPsych? Do you also publish separate packages as a contributor or do you do pull requests into their main repository?

**Comment by younesStrittmatter** [comment]
jsPsych consists of a core repository, maintained by the core team, and a contrib repository, where users can contribute their own plugins and extensions. Anyone can contribute to both repositories, but the core team is responsible for maintaining and publishing the core repository, while contributions to the contrib repository are reviewed by the team but are not necessarily maintained or updated regularly.
The publishing for both repositories is not done by the contributor but by the jsPscyh-team. I think, that each contribution in the jsPsych-contrib repository is published as an individual npm package. So they have one repository containing a lot of individual npm packages (while we have a repository for each pip package).

**Comment by younesStrittmatter** [comment]
So, in short: the contribution gets pulled into a main repository but is still published as an individual package.

**Comment by younesStrittmatter** [comment]
I think namespacing the contributions also has the advantage that it clearly distinguishes between "core" and thoroughly reviewed packages, and outside contributions, while making it easy to contribute. When the user imports something from "contrib" it is clear, that this is on their own risk and it might break other code.

**Comment by musslick** [comment]
We should discuss having a separate repo for outside contributions then. I think that nicely complements the more rigorous sub-package solution. We could also run both and see which is most adapted.

**Reply by younesStrittmatter** [reply]
I don't think a separate repo makes sense since all the packages are individual repos anyway. I meant adding a contrib folder to the structure. So instead of contributors adding their packages to src/autora/theorist/new_theorist, they would have the option to add to src/autora/contrib/theorist/new_theorist. We would still review these packages before adding them to the autora-mother repo as an optional dependency, but the barriers could be lower than adding a package published under src/autora/theorist/new_theorist directly.

**Reply by younesStrittmatter** [reply]
At least, I think this would be the easier way to go about it than adding an additional package called autora-contrib and managing stuff there (and then adding autora-contrib to the autora-mother repo as an optional dependency). For me, that just seems like an unnecessary extra step (but I might well be mistaken)

**Reply by musslick** [reply]
Wouldn't that encounter the problem of having to manage conflicting dependencies between contributions (conflicts we wanted to avoid with separate packages)?

**Reply by younesStrittmatter** [reply]
The autora-mother package and the autora-contrib packages would be the same in the sense that they are just “organizing” stuff (especially documentation) without actual code. So the structure stays the same, and the packages are still separate. The autora-contrib repo would basically be a list of “approved” contrib packages (as the autor-mother package is basically a list of "approved" packages), but I think we should just add that to the autora-mother repo without the unnecessary extra repo.

**Reply by musslick** [reply]
I'm not sure I understand "The autora-contrib repo would basically be a list of “approved” contrib packages (as the autor-mother package is basically a list of "approved" packages": If both have approved packages, what is the difference? That the autora-mother package only has internal contributions?

I also don't understand what contributing to the contrib-package would entail. Would you still go through the process of releasing separate a sub-package to manage depenency conflicts? If so, why would this be easier than contributing a subpackage to the autora mother repo?

**Reply by younesStrittmatter** [reply]
I wouldn't add a repo (just a namespace), which would be src/autora/contrib/.... instead of src/autora/.... and the only difference is, that our review process is more lax. The publishing is still the same and the packages are still independent, they just are not directly under the autora namespace.
The benefit is this:
When the user installs a package that is in the contrib namespace, they clearly see that it is a contribution and not maintained by the core team. And when importing, they see that again. So there is no confusion between what is "core" to the project (aka developed by us or approved by us)
So there would be "three different" stages of packages (and install/import methods):

-----
1. Free, unhinged contribution (this can be done anyway, and we can not prevent it, but we can warn users from installing like this)

installation: pip install autora-theorist-malware-theorist
import: from autora.theorist.malware.theorist import TrojanRegressor

If the user installs like this, that is their own risk. The package could be approved or even core, but there is no guarantee. It could also be something completely different. But this is still usefull if for example, a lab wants to create functionality that is only relevant for themselves and easily share it between team members. This could also be contributions, that we don't approve or even know of, but they can still be useful. 

-----
2. Contributions that have been seen and approved but are not core

installation: pip install autora["contrib-theorist-broken-theorist"]
import: from autora.contrib.theorist.broken_theorist import BrokenRegressor

These packages have been seen and approved by the team at least once and they have been working with autora-core at least once, but we don't guarantee full documentation and maintenance. Also, we can't guarantee that these packages work well with others (btw. the documentation on what versions/dependencies exist might be lacking). 

---
3. Core packages and packages that were reviewed thoroughly.

installation: pip install autora["theorist-awesome-theorist"]
import: from autora.theorist.awesome_theorist import AwesomeRegressor

These packages have been approved or developed by the team, they are maintained and well-documented.

**Reply by musslick** [reply]
Hmm.... One more clarification: We currently refer to "core" as the core-autora package which does not include any other packages. Here, by "core packages", do you mean packages integrated into the "autora" mother package, not the "autora-core" package?

**Reply by younesStrittmatter** [reply]
Yes, exactly. Sorry for the confusion.

**Reply by musslick** [reply]
Ok, got it. That sounds good. As for contribution version 3: How would we ensure to review/control the content of external sub-packages that are not authored by us? I assume we could encourage contact the authors separately and provide instructions for requirements but, we can't formally approve (e.g., by means of PRs) by default, right?

**Reply by younesStrittmatter** [reply]
I am not sure if I understand correctly, but the contributions only controlled by the mother package and therefore the means to install the packages. If somebody wants to use our template to create a pip package under autora/theorist/malware, we can not prevent that. But we are only adding packages that we have approved to the autora mother package. Therefore the installation via autora["..."] is controlled by us and "save" (@hollandjg: That is correct, right?). So the control is in the mother package, which only contains approved packages either under autora/contrib or directly under autora/

**Reply by musslick** [reply]
I thought so differently. That is, we (as core team) also publish separate packages and treat them as external contributions (e.g., autora-theorist-darts). The only thing we control is whether we include the sub-package as a dependency in the autora "mother" package.

**Reply by younesStrittmatter** [reply]
You are still right. That is still the case.

**Reply by younesStrittmatter** [reply]
I'm sorry for the confusion. The mother package stays the same, but has an additional namespace contrib where outside packages are added as optional dependencies (the namespace contrib just signifies that this is an okay contribution but not something like autora-theorist-darts or autora-core which is developed by the AutoRA team itself)

**Reply by musslick** [reply]
Ah, so we would not review those and not add them as dependencies?  I think I understand now, and it does make sense. Although now, I am worried that the contributors will get easily confused, about which contribution option to choose, so we should either evaluate whether this confusion is worth the functionality and/or make our best effort to communicate these different types of contributions to the community.
